src/main/java/com/vmware/vipclient/i18n/messages/service/ProductService.java

```diff
@@ -72,44 +73,75 @@ public ProductService(BaseDTO dto) {
+        return getSupportedLocales(false);
+    }
+
     /**
-     * Retrieves the list of locales of a product. It recursively applies data source fallback mechanism in case of failure.
+     * Retrieves the set of supported locales.
      *
-     * @return list of locales of the product specified in the dto object, or an empty list in case of failure to retrieve from any data source.
+     * @param refreshCache If true, it will trigger a cache populate or refresh as necessary before returning.
+     *                         If false, it will return the data from the cache as is, or an empty Set if not in cache.
+     * @return The set of supported locales.
      */
-    public Set<Locale> getSupportedLocales() {
-        return langTagtoLocaleSet(getSupportedLanguageTags());
-    }
-
-    public Set<String> getSupportedLanguageTags() {
+    private Set<String> getSupportedLocales(boolean refreshCache) {
         Iterator<DataSourceEnum> msgSourceQueueIter = VIPCfg.getInstance().getMsgOriginsQueue().iterator();
         Set<String> supportedLangTags = new HashSet<>();
         while(msgSourceQueueIter.hasNext()) {
-            supportedLangTags.addAll(getSupportedLanguageTags(msgSourceQueueIter.next()));
+            if (refreshCache)
+                supportedLangTags.addAll(getSupportedLocales(msgSourceQueueIter.next()));
+            else
+                supportedLangTags.addAll(getCachedSupportedLocales(msgSourceQueueIter.next()));
         }
         return supportedLangTags;
     }
 
-    public Set<String> getSupportedLanguageTags(DataSourceEnum dataSource) {
+    /**
+     * Retrieves the cached set of locales that are supported in the given data source.
+     *
+     * @param dataSource The data source
+     * @return The cached set of locales that are supported in the given data source.
+     */
+    public Set<String> getCachedSupportedLocales(DataSourceEnum dataSource) {
+        CacheService cs = new CacheService(new MessagesDTO(dto));
+        MessageCacheItem cacheItem = cs.getCacheOfLocales(dataSource);
+        return cacheItem == null ? new HashSet<>() : cacheItem.getCachedData().keySet();
+    }
+
+    /**
+     * Retrieves the set of locales that are supported in the given data source.
+     * It will trigger a cache populate or refresh as necessary before returning.
+     *
+     * @param dataSource The data source
+     * @return The set of locales supported in the given data source.
+     */
+    public Set<String> getSupportedLocales(DataSourceEnum dataSource) {
         CacheService cs = new CacheService(new MessagesDTO(dto));
         MessageCacheItem cacheItem = cs.getCacheOfLocales(dataSource);
         if (cacheItem != null) {
-            if (cacheItem.isExpired())
-                refreshLocalesCacheItemTask(cacheItem, dataSource);
+            if (cacheItem.isExpired()) {
+                synchronized (cacheItem) { // Allow only 1 thread to refresh the cacheItem at a time.
```

Comment:
Yes, and that is why inside the synchronized block, another `if (cacheItem.isExpired())` is done so that it will not proceed with the update if previous thread has already refreshed the cacheItem.

Reply:
Anyway, I have removed this part out of this PR to narrow down the scope. I will include it in another PR.